napari/_app_model/_app.py

```diff
@@ -25,22 +55,162 @@ def __init__(self, app_name=APP_NAME) -> None:
+        # "prefix:suffix", while app-model supports extra levels of nesting
+        # e.g.
+        #     app-model allows "napari:viewer:toggle_theme"
+        #     action_manager wants "napari:toggle_theme"
+        # so this hack works if we keep the prefix and suffix the same when
+        # porting to app-model
+        prefix, *_group, command = action.id.split(":")
+
+        if isinstance(action.callback, GeneratorCallback):
+
+            def _callback(*args, **kwargs):
+                self.get_app().commands.execute_command(action.id).result()
+                yield
+                self.get_app().commands.execute_command(action.id).result()
+
+        else:
+
+            def _callback(*args, **kwargs):
+                return (
+                    self.get_app().commands.execute_command(action.id).result()
+                )
+
+        _callback.__name__ = command
+
+        action_manager.register_action(
+            name=f"{prefix}:{command}",
+            command=_callback,
+            description=action.title,
+            keymapprovider=keymapprovider,
+            repeatable=isinstance(action, RepeatableAction),
+        )
+
     @classmethod
     def get_app(cls, app_name: str = APP_NAME) -> NapariApplication:
         return Application.get_app(app_name) or cls()  # type: ignore[return-value]
 
+    def register_action(self, action: Action) -> Callable[[], None]:
+        dispose = super().register_action(action)
+        if isinstance(action, RepeatableAction) and action.repeatable:
+            self.set_action_repeatable(action.id, True)
+
+            def _dispose():
+                dispose()
+                self._repeatable_actions.remove(action.id)
+
+            return _dispose
+        return dispose
+
+    def set_action_repeatable(self, action_id: str, repeatable: bool):
+        """Registers an action as repeatable or not.
+
+        Parameters
+        ----------
+        action_id : str
+            Unique identifier of the action.
+        repeatable : bool
+            Whether or not to set the action as repeatable.
+        """
+        if action_id not in self.commands:
+            raise ValueError(f'Command {id!r} not registered')
+
+        if repeatable:
+            self._repeatable_actions.add(action_id)
+        else:
+            self._repeatable_actions.discard(action_id)
+
+    def action_is_repeatable(self, action_id: str) -> bool:
+        """Determines if an action is repeatable.
+
+        Parameters
+        ----------
+        action_id : str
+            Unique identifier of the action.
+
+        Returns
+        -------
+        bool
+            Whether the action is repeatable or not.
+        """
+        return action_id in self._repeatable_actions
+
+    def _on_shortcuts_changed(self, _):
+        self.keybindings.discard_entries(KeyBindingWeights.USER)
+        with warnings.catch_warnings():
+            warnings.simplefilter('ignore', UserWarning)
```

Comment:
Why these lines are needed? because of the bug in `warnings.catch_warnings()` clean register of already emitted warnings. So causes spawning obsolete warnings in GUI.

Reply:
this is because setting a modifier-only key binding will generate a `UserWarning` if they used e.g. `KeyCode.Ctrl` instead of `KeyCode.CtrlCmd`. are you saying that this is already suppressed elsewhere?

Reply:
No. The problem is in this cPython issue https://github.com/python/cpython/issues/73858

Reply:
well that's dumb. I'll just add a parameter to suppress warnings then